File: server/app/services/task/tools/web_search.py
"""
Web search tool for retrieving information from the internet.

Supports multiple search providers:
- Tavily: Requires API key, high quality results
- DuckDuckGo: No API key required, but may be blocked in some networks
"""

import json
from typing import Any, Dict, List, Optional

from app.services.task.tools.base import Tool
from app.logger import logger


class WebSearchTool(Tool):
    """
    Tool for searching the web.

    Uses the configured search provider (Tavily or DuckDuckGo).
    Returns a list of search results with title, URL, and snippet.
    """

    DEFAULT_NUM_RESULTS = 5

    def __init__(self, search_config: Optional[Any] = None):
        """
        Initialize the web search tool.

        Args:
            search_config: SearchConfig instance containing provider and api_key.
                          If None, the tool will not be functional.
        """
        self._search_config = search_config

    @property
    def name(self) -> str:
        return "web_search"

    @property
    def schema(self) -> Dict[str, Any]:
        return {
            "type": "function",
            "function": {
                "name": "web_search",
                "description": (
                    "Search the web for information. Use this tool to find "
                    "current information, documentation, or answers to questions."
                ),
                "parameters": {
                    "type": "object",
                    "properties": {
                        "query": {
                            "type": "string",
                            "description": "The search query",
                        },
                        "num_results": {
                            "type": "integer",
                            "description": "Number of results to return (default: 5)",
                            "default": self.DEFAULT_NUM_RESULTS,
                        },
                    },
                    "required": ["query"],
                },
            },
        }

    async def execute(self, **kwargs) -> str:
        """
        Execute a web search and return results.

        Args:
            query: The search query string.
            num_results: Number of results to return (default: 5).

        Returns:
            JSON string with a list of search results.
        """
        query = kwargs.get("query", "")
        num_results = kwargs.get("num_results", self.DEFAULT_NUM_RESULTS)

        if not query:
            return json.dumps({"results": [], "error": "Empty query"})

        if not self._search_config:
            return json.dumps({
                "results": [],
                "error": "Search engine not configured"
            })

        provider = self._search_config.provider
        api_key = self._search_config.api_key

        logger.info(
            f"WebSearchTool searching: query='{query[:50]}...', "
            f"provider={provider}, num_results={num_results}"
        )

        try:
            if provider == "tavily":
                results = await self._search_tavily(api_key, query, num_results)
            elif provider == "duckduckgo":
                results = await self._search_duckduckgo(query, num_results)
            else:
                return json.dumps({
                    "results": [],
                    "error": f"Unknown search provider: {provider}"
                })

            logger.info(f"WebSearchTool found {len(results)} results")
            return json.dumps({"results": results}, ensure_ascii=False)

        except Exception as e:
            logger.error(f"WebSearchTool error: {str(e)}", exc_info=True)
            return json.dumps({"results": [], "error": str(e)})

    async def _search_tavily(
        self, api_key: str, query: str, num_results: int
    ) -> List[Dict[str, str]]:
        """
        Perform a Tavily search.

        Args:
            api_key: Tavily API key.
            query: Search query string.
            num_results: Maximum number of results to return.

        Returns:
            List of dicts with title, url, and snippet keys.
        """
        from tavily import TavilyClient

        client = TavilyClient(api_key=api_key)
        result = client.search(query=query, max_results=num_results)

        results = []
        for item in result.get("results", []):
            results.append({
                "title": item.get("title", ""),
                "url": item.get("url", ""),
                "snippet": item.get("content", ""),
            })

        return results

    async def _search_duckduckgo(
        self, query: str, num_results: int
    ) -> List[Dict[str, str]]:
        """
        Perform a DuckDuckGo search.

        Args:
            query: Search query string.
            num_results: Maximum number of results to return.

        Returns:
            List of dicts with title, url, and snippet keys.
        """
        from duckduckgo_search import DDGS

        results = []
        with DDGS() as ddgs:
            search_results = list(ddgs.text(query, max_results=num_results))

        for item in search_results:
            results.append({
                "title": item.get("title", ""),
                "url": item.get("href", ""),
                "snippet": item.get("body", ""),
            })

        return results

File: server/internal/service/task/tools/web_search.go
package tools

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"

	"private-buddy-server/internal/model"
	"private-buddy-server/internal/service/llm"
)

// WebSearchTool searches the web for information using configured search providers.
// Currently supports Tavily as the search provider.
// Returns a list of search results with title, URL, and snippet.
type WebSearchTool struct {
	searchConfig *model.SearchConfig // Search configuration containing provider and API key
}

// NewWebSearchTool creates a WebSearchTool with the given search configuration.
func NewWebSearchTool(searchConfig *model.SearchConfig) *WebSearchTool {
	return &WebSearchTool{searchConfig: searchConfig}
}

func (w *WebSearchTool) Name() string { return "web_search" }

func (w *WebSearchTool) Schema() llm.FunctionDefinition {
	return llm.FunctionDefinition{
		Name:        "web_search",
		Description: "Search the web for information. Use this tool to find current information, documentation, or answers to questions.",
		Parameters: map[string]interface{}{
			"type": "object",
			"properties": map[string]interface{}{
				"query": map[string]interface{}{
					"type":        "string",
					"description": "The search query",
				},
				"num_results": map[string]interface{}{
					"type":        "integer",
					"description": "Number of results to return (default: 5)",
					"default":     5,
				},
			},
			"required": []string{"query"},
		},
	}
}

// SearchResult represents a single web search result.
type SearchResult struct {
	Title   string `json:"title"`
	URL     string `json:"url"`
	Snippet string `json:"snippet"`
}

// SearchResponse is the structured response for web search execution.
type SearchResponse struct {
	Results []SearchResult `json:"results"`
	Error   string         `json:"error,omitempty"`
}

// Execute performs a web search and returns results as JSON.
// Returns error message if search config is not available or provider is unknown.
func (w *WebSearchTool) Execute(args map[string]interface{}) (string, error) {
	query, _ := args["query"].(string)
	numResults := 5
	if n, ok := args["num_results"].(float64); ok {
		numResults = int(n)
	}

	if query == "" {
		resp, _ := json.Marshal(SearchResponse{Error: "Empty query"})
		return string(resp), nil
	}

	if w.searchConfig == nil || !w.searchConfig.IsAvailable() {
		resp, _ := json.Marshal(SearchResponse{Error: "Search engine not configured"})
		return string(resp), nil
	}

	provider := w.searchConfig.Provider
	apiKey := w.searchConfig.APIKey

	var results []SearchResult
	var err error

	switch provider {
	case "tavily":
		results, err = w.searchTavily(apiKey, query, numResults)
	default:
		resp, _ := json.Marshal(SearchResponse{Error: fmt.Sprintf("Unknown search provider: %s", provider)})
		return string(resp), nil
	}

	if err != nil {
		resp, _ := json.Marshal(SearchResponse{Error: err.Error()})
		return string(resp), nil
	}

	resp, _ := json.Marshal(SearchResponse{Results: results})
	return string(resp), nil
}

// searchTavily performs a search using the Tavily API.
// Sends a POST request to the Tavily search endpoint and parses the response.
func (w *WebSearchTool) searchTavily(apiKey, query string, numResults int) ([]SearchResult, error) {
	payload := map[string]interface{}{
		"api_key":     apiKey,
		"query":       query,
		"max_results": numResults,
	}
	body, _ := json.Marshal(payload)

	resp, err := http.Post("https://api.tavily.com/search", "application/json", bytes.NewReader(body))
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	data, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}

	var tavilyResp struct {
		Results []struct {
			Title   string `json:"title"`
			URL     string `json:"url"`
			Content string `json:"content"`
		} `json:"results"`
	}
	if err := json.Unmarshal(data, &tavilyResp); err != nil {
		return nil, err
	}

	results := make([]SearchResult, 0, len(tavilyResp.Results))
	for _, item := range tavilyResp.Results {
		results = append(results, SearchResult{
			Title:   item.Title,
			URL:     item.URL,
			Snippet: item.Content,
		})
	}
	return results, nil
}
